File: app/src/main/java/netpro/keytransmitter/KeyRecyclerViewAdapter.java
package netpro.keytransmitter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.LinkedList;
import java.util.List;

public class KeyRecyclerViewAdapter extends RecyclerView.Adapter<KeyViewHolder> {
    private List<Key> datasource = new LinkedList<>();
    private Context context;


    public KeyRecyclerViewAdapter(Context context) {
        this.context = context;
    }

    @Override
    public KeyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_cardview, parent, false);
        return new KeyViewHolder(v);
    }

    @Override
    public void onBindViewHolder(KeyViewHolder view, int position) {
        Key key = datasource.get(position);
        view.setKey(key);
    }

    @Override
    public int getItemCount() {
        return datasource.size();
    }

    public void addView(Key key) {
        datasource.add(key);
        notifyItemInserted(getItemCount() - 1);
    }

    public void addAllView(List<Key> list) {
        int preSize = getItemCount();
        datasource.addAll(list);
        notifyItemRangeInserted(preSize, list.size() - 1);
    }

    public Key get(int index) {
        return datasource.get(index);
    }

    public void setDatasource(List<Key> datasource) {
        this.datasource = datasource;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/netpro/keytransmitter/KeyRecyclerViewAdapter.kt
package netpro.keytransmitter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import java.util.*

/**
 * MainActivityのRecyclerViewの
 */
class KeyRecyclerViewAdapter : RecyclerView.Adapter<KeyViewHolder>() {
    var keyList: MutableList<Key> = LinkedList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): KeyViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.layout_cardview, parent, false)
        return KeyViewHolder(v)
    }

    override fun onBindViewHolder(view: KeyViewHolder, position: Int) {
        val key = keyList[position]
        view.setKey(key)
    }

    override fun getItemCount(): Int {
        return keyList.size
    }

    fun addView(key: Key) {
        keyList.add(key)
        notifyItemInserted(itemCount - 1)
    }

    fun addAllView(list: List<Key>) {
        val preSize = itemCount
        keyList.addAll(list)
        notifyItemRangeInserted(preSize, list.size - 1)
    }

    operator fun get(index: Int): Key {
        return keyList[index]
    }
}